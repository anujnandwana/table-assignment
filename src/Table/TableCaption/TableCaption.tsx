import React from "react";
import { TableData } from "../../data/data";
import DownloadAction from "../Actions/Download";
import SelectAll from "../Actions/SelectAll";
import SelectedText from "../Actions/SelectedText";

interface TableCaptionProps {
  selectedRowIds: string[];
  setSelectedRowIds: (ids: string[]) => void;
  selectAll: boolean;
  tableData: TableData[];
}

const TableCaption: React.FC<TableCaptionProps> = ({
  selectedRowIds,
  setSelectedRowIds,
  selectAll,
  tableData,
}) => {
  const [downloadAbleData, setDownloadAbleData] = React.useState<TableData[]>([]);

  
  /**
   * Update the disable/enable status of download btn by checking available status
   */
  React.useEffect(() => {
    const availableStatusData = tableData.filter(
      (row) =>
        selectedRowIds.includes(row.id as string) && row.status === "available"
    );
    setDownloadAbleData(availableStatusData);
  }, [tableData, selectedRowIds]);


  return (
    <caption>
      <SelectAll selectAll={selectAll} setSelectedRowIds={setSelectedRowIds} tableData={tableData} />
      <SelectedText selectedRowIds={selectedRowIds} />
      <DownloadAction downloadAbleData={downloadAbleData} />
    </caption>
  );
};

export default TableCaption;
