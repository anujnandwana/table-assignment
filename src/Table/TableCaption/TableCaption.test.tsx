import TableCaption from './TableCaption';
import { mount, shallow } from 'enzyme';
import { TableData } from '../../data/data';

describe('Table component', () => {
  const mockData: TableData[] = [
    { id: "mockedId", device: 'Device 1', path: '/path/1', name: 'jui', status: 'available' }
  ];

  const selectedRowIds = ["mockedId"];
  const setSelectedRowIds = jest.fn()
  const selectAll = true;
  const tableData = mockData;

  test('TableCaption component should match snapshot', () => {
    const component = shallow(<TableCaption selectedRowIds={selectedRowIds} setSelectedRowIds={setSelectedRowIds} selectAll={selectAll} tableData={tableData} />);
    expect(component).toMatchSnapshot();
  });

  test('selectAllCheckbox should be present with checked true if all is selected', () => {
    const component = mount(<TableCaption selectedRowIds={selectedRowIds} setSelectedRowIds={setSelectedRowIds} selectAll={selectAll} tableData={tableData} />);
    const selectAllCheckbox = component.find("input");

    expect(selectAllCheckbox.length).toBe(1);
    expect(selectAllCheckbox.at(0).prop('checked')).toBeTruthy();
  });

  test('selectAllCheckbox should be present with checked false if all is not selected', () => {
    const component = mount(<TableCaption selectedRowIds={selectedRowIds} setSelectedRowIds={setSelectedRowIds} selectAll={false} tableData={tableData} />);
    const selectAllCheckbox = component.find("input");

    expect(selectAllCheckbox.length).toBe(1);
    expect(selectAllCheckbox.at(0).prop('checked')).toBeFalsy();
  });

  test('select count should show Selected 1 when 1 row is selected ', () => {
    const component = mount(<TableCaption selectedRowIds={selectedRowIds} setSelectedRowIds={setSelectedRowIds} selectAll={false} tableData={tableData} />);    
    const selectedText = component.find("span.selected-text");

    expect(selectedText.at(0).text()).toBe("Selected 1");
  });

  test('select count should show None Selected when 0 row is selected ', () => {
    const component = mount(<TableCaption selectedRowIds={[]} setSelectedRowIds={setSelectedRowIds} selectAll={false} tableData={tableData} />);    
    const selectedText = component.find("span.selected-text");

    expect(selectedText.at(0).text()).toBe("None Selected");
  });


  test('Download btn should be enabled when available status row is selected ', () => {
    const component = mount(<TableCaption selectedRowIds={selectedRowIds} setSelectedRowIds={setSelectedRowIds} selectAll={false} tableData={tableData} />);    
    const downloadBtn = component.find("button.active");

    expect(downloadBtn.at(0).prop("disabled")).toBeFalsy();
  });

  test('Download btn should be disabled when available status row is selected ', () => {
    const component = mount(<TableCaption selectedRowIds={["abc"]} setSelectedRowIds={setSelectedRowIds} selectAll={false} tableData={tableData} />);    
    const downloadBtn = component.find("button");

    expect(downloadBtn.at(0).prop("disabled")).toBeTruthy();
  });

});
