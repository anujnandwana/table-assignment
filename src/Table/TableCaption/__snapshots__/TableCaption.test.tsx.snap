// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component TableCaption component should match snapshot 1`] = `ShallowWrapper {}`;
